pkg/codegen/hcl2/model/expression.go

```diff
@@ -1034,7 +1034,13 @@ func (x *FunctionCallExpression) Typecheck(typecheckOperands bool) hcl.Diagnosti
 	typecheckDiags := typecheckArgs(rng, x.Signature, x.Args...)
 	diagnostics = append(diagnostics, typecheckDiags...)
 
-	x.Signature.ReturnType = liftOperationType(x.Signature.ReturnType, x.Args...)
+	// Unless the function is already automatically using an
+	// Output-returning version, modify the signature to account
+	// for automatic lifting to Promise or Output.
+	_, isOutput := x.Signature.ReturnType.(*OutputType)
+	if !isOutput {
+		x.Signature.ReturnType = liftOperationType(x.Signature.ReturnType, x.Args...)
+	}
```

Comment:
This is interesting. I think that the lifting behavior is wrong in general here: essentially, it's lifting based on the assumption that all parameter types are prompt, so any eventual inputs must transform eventual outputs. Instead, the lifting should be conditional on the argument type not being assignable to the parameter type. In other words, a call that passes an `Output<T>` arg to an `Output<T>` parameter does not require lifting, but a call that passes an `Output<T>` arg to a `T` parameter does.

Reply:
Filed https://github.com/pulumi/pulumi/issues/8439 to track.

Reply:
I'm looking at it now briefly now.

Tried this but it fails:

```go

	diags := typecheckArgs(rng, x.Signature, x.Args...)
	lifted := liftOperationType(x.Signature.ReturnType, x.Args...)

	if !lifted.equals(x.Signature.ReturnType, nil) {
		fmt.Printf("LIFTED BUT DIAGS: %v\n", diags)
	}

	if diags.HasErrors() {
		x.Signature.ReturnType = lifted
	}

```

```
=== RUN   TestGenerateProgram/AWS_EKS
LIFTED BUT DIAGS: no diagnostics
```

Reply:
It's super tricky, `x.Parameters[0].Type == Dynamic` in the failing tests. So we don't know the real parameter type at this point. I'll need help on  #8439 later on, I'm also confused why this code is modifying the signature. It's kind of surprising, we'll need to put some comments and point to the overview of the Typecheck pass and how it interplays with other passes I think, to make this easy to understand.